### ISSUE TITLE
handle vue dependencies in nuxt 3 with externals or 'smart' transpilation

### ISSUE THREAD

**Issue by danielroe** [nuxt/framework#4084]
## Summary

At the moment, in both Bridge & Nuxt 3, we transpile `vue`. This decreases the space taken up in our server builds, and also protects against multiple instances of Vue being present in development (or even in build, due to ESM resolution issues).

However, it comes with the issue that if the end user adds any libraries or other files (outside of _their_ project) that import vue, we now have multiple copies of Vue present - both the version that has been bundled and an external version in `node_modules`.

It also means that it is more difficult for users to pick the version of Vue they wish to use in their project (for example, if they want a version with a runtime compiler).

## Possible solutions

1. Users can manually add any such libraries to `build.transpile` - this is our current recommendation.
2. Building on 1, we can also use vite's dependency graph to detect nested dependencies of vue and recommend adding them to `build.transpile`
3. We can fully externalise `vue`, at the cost of greater disk space and the danger of resolving multiple vue versions depending on the ESM status of dependencies.
4. Building on 3, we can also improve our NFT support by adding 'soft-transpilation' to optimize/tree-shake externals.

## Related issues

* https://github.com/nuxt/framework/issues/3817
* https://github.com/nuxt/framework/issues/2530
* https://github.com/nuxt/framework/issues/2326#issuecomment-1026081414
* https://github.com/nuxt/framework/issues/3688
* https://github.com/nuxt/framework/issues/3162
* https://github.com/nuxt/framework/issues/2682

**Comment by zefman** [comment]
Hi @danielroe should adding the library to `build.transpile` always fix the issue? I have a custom component library setup with pnpm alongside Nuxt in a mono repo and still face issues even after adding to the transpile array. The component library is very simple using vite to bundle with vue externalised etc.

**Comment by dakt** [comment]
We have same issue in our project  with dev erroring out:

```
[Vue warn]: inject() can only be used inside setup() or functional components.
[Vue warn]: inject() can only be used inside setup() or functional components.
[Vue warn]: inject() can only be used inside setup() or functional components.
[Vue warn]: Unhandled error during execution of setup function
  at <RouterLink to="typography" activeClass=undefined exactActiveClass=undefined  ... >
[nuxt] [request error] Cannot destructure property 'options' of 'vue.inject(...)' as it is undefined.
```

This is only when we slot <Nuxt* /> components through our UI lib like this:

```html
<UIProvider theme="foo">
  <NuxtLink to="foobar">Foobar</NuxtLink>
</UIProvider>
```

but works fine in this scenario:

```html
<UIProvider theme="foo">
  <h1>Hello World</h1>
</UIProvider>
```

`build.transpile` does nothing for us since it's probably not working in monorepo for some reason.

**Comment by zefman** [comment]
@dakt Interesting you are in a monorepo too, it does sound like build.transpile doesn't work in monorepos. 

I would be really interested if you find a solution! This is a pretty big blocker for us at the moment and is ruining our otherwise lovely monorepo setup haha

**Comment by dakt** [comment]
@zefman 

Didn't spend much time trying to figure out why `built.transpile` want't working, but we managed to make the project which is using our UI lib to work. We just did whole monorepo cleanup (remove all build/dev artefacts) and used `pnpm --shamefully-hoist' to install Nuxt dependencies.  This caused all Nuxt's deps to bubble up flat in node_modules and probably allowed  implicit dependencies to be resolved / imported.

**Comment by zipme** [comment]
hi @danielroe , it still happens with 
```ts
experimental: {
  externalVue: true,
},
```

**Comment by zipme** [comment]
@danielroe oh sorry, I was meant to comment on this issue https://github.com/nuxt/framework/issues/2530. It turned out I have different vue version set in different packages, once I unified them the problem went away.

**Comment by Lilja** [comment]
@danielroe I was able to add `experimental: { externalVue: true }` to the reproduction in #2530, still got the same issues.

**Comment by P4sca1** [comment]
@danielroe Were there any recent changes regarding this topic? Because we noticed, that you are required to add `@headlessui/vue` to `build.transpile` again.
More context: https://github.com/P4sca1/nuxt-headlessui/issues/13

**Comment by danielroe** [comment]
No, there _should_ not be a change. There might have been a change in vite or nitro bundling though, so if you are happy to raise a new issue, we can look to resolve it.

**Comment by BakoviDagi** [comment]
I'm also having problems importing a Vue 3 component library that uses composition API.

I tried adding the component to `build.transpile` and setting `experimental.externalVue` and it's still not working

![image](https://user-images.githubusercontent.com/56078237/204552270-37a703ea-061d-4b51-86fd-d310d2f8b0a9.png)

**Comment by BakoviDagi** [comment]
This was fixed for me by removing all of the `import { ... } from 'vue'` from my library. Are we not supposed to import things from vue?
